### PR TITLE
Illustrate adding a wallet_receive function to the default template project

### PR DESCRIPTION
**Pull request by lsgunnlsgunn**
WIP to try out adding a wallet_receive function to a Motoko program to work with the Rust wallet created by default for local development. I'm trying to test how someone might add the wallet-related functions using our quick start demo project. 

### Demo

1.  Clone the project.
2. Change to the `hello_cycles` directory.
3. Run `dfx start --clean --background`.
4. Run `dfx deploy --argument '(20000000000000)'`. The argument is required because this is an actor class.
  Note the canister identifier for the `hello_cycles` canister.
5. Run `dfx canister call hello_cycles greet '("everyone")'` to verify the default function works as expected.
6. Run `more .dfx/local/wallets.json` to find the wallet canister identifier.
7. Run a command similar to the following to try to send cycles to the `hello_cycles` canister from the local wallet.
  `dfx canister call rwlgt-iiaaa-aaaaa-aaaaa-cai wallet_send '(record { canister = principal "rrkah-fqaaa-aaaaa-aaaaq-cai"; amount = (2000000000000:nat64); } )'`

  `rwlgt-iiaaa-aaaaa-aaaaa-cai` is the wallet canister for the active identity created automatically.
  `rrkah-fqaaa-aaaaa-aaaaq-cai` is the hello_cycles canister with the test `wallet_receive` function.

  Currently, this returns an error like this:
  `The Replica returned an error: code 5, message: "Canister rwlgt-iiaaa-aaaaa-aaaaa-cai trapped explicitly: Custom("No more values to deserialize")"`
  Possibly because the wallet canister expects an event?

### CODE REVIEW

**Comment by crusso** [comment]
I took a quick look, but didn't try to test my proposal.

My hunch is that the rust wallet is failing to deserialize the result of its internal call to Motoko `wallet_receive`.

If that is meant to have the same signature as the wallet's own `wallet_receive `method, then the Motoko `wallet_receive `should be returning a Motoko` { accepted : Nat64 }  `(Candid record { accepted : nat64; }` (I guess)).

In your code, Motoko`wallet_receive` has type `() -> async ()`, so is returning no values (I think). 

Careful, Motoko's `accept` function returns a Nat, not Nat64, so you'll need to do a conversion via 'Nat64.fromNat(n)' before returning the record.

If this doesn't work for you, let me know and I'll try myself tomorrow (or just tell me you didn't get around to it).

**Comment by crusso** [comment]
#72 contains the fix as a diff to your branch. 

Seems to work, without errror, but I can't call `hello_cycles.balance()` from the `dfx` because the assert fails for some reason.

**Comment by lsgunnlsgunn** [comment]
That's a good question. I can see my current identity and get the principal for that, but there isn't currently a `get-controller` (AFAIK) command to see the current controller for a canister. I'll file an issue for that.

**Comment by lsgunnlsgunn** [comment]
I removed the `assert` and my little example works:
```
dfx deploy --argument '(120000000000000)'
... Deployed canisters.

dfx canister call hello_cycles greet '("everyone")'
("Hello, everyone!")

dfx canister call hello_cycles wallet_balance
(0)

dfx canister call rwlgt-iiaaa-aaaaa-aaaaa-cai wallet_send '(record { canister = principal "rrkah-fqaaa-aaaaa-aaaaq-cai"; amount = (2000000000000:nat64); } )'
()

```
dfx canister call hello_cycles wallet_balance
(2_000_000_000_000)

**Comment by crusso** [comment]
Glad to it see it works. 
I wonder, when you install the canister, does the wallet become the caller of the install method? That might explain why the assert failed. Maybe add a method to display the textual version of owner and msg.caller, so you can see which it is? Better yet, ask Hans...

File: motoko/hello_cycles/src/hello_cycles/main.mo [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  var balance = 0;
+
+  public shared(msg) func wallet_balance() : async Nat {
+    return balance;
+  };
+
+  public func wallet_receive() : async { accepted: Nat64 } {
+    let amount = Cycles.available();
+    let limit = capacity - balance;
+    let accepted =
```
Comment (crusso) [inline motoko/hello_cycles/src/hello_cycles/main.mo:19]: maybe rename `accepted` -> `spare` and `rec_accepted` -> `accepted`

**Comment by crusso** [comment]
Might be nice to implement `wallet_send` too, so you can illustrate `Cycles.add()` and `Cycles.refunded()`.

**Comment by lsgunnlsgunn** [comment]
Yep, it would be cool to either extend this example or add a second version with the `wallet_send`, `Cycles.add()`, and `Cycles.refunded()` operations.
 I thought this was a good starting point because people will get the Rust wallet automatically but then might need to add the receive function to their existing canisters that aren't wallets.

I really appreciate your help in getting these little examples working!

File: motoko/hello_cycles/src/hello_cycles/main.mo [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+  let owner = msg.caller;
+
+  var balance = 0;
+
+  public shared(msg) func wallet_balance() : async Nat {
+    return balance;
+  };
+
+  public func wallet_receive() : async { spare: Nat64 } {
```
[the remark this answers is not in the document]
Comment (crusso) [reply motoko/hello_cycles/src/hello_cycles/main.mo:16]: Sorry, I think I was a bit confusing. 

Maybe it's jut me, but I think it makes more sense to call and report the amount `accepted` - `spare `is how much you could `accept` without exceeding capacity. 

Does that make sense?

Also, the rust wallet API expects the returned record field to be called `accepted`, I think, though the code currently ignores the result on send AFAICT.

See here:
https://github.com/dfinity/wallet-rs/blob/48881d6c41e95b8796aa77bd3601209a15482979/wallet/src/lib.rs#L252
and here:
https://github.com/dfinity/wallet-rs/blob/48881d6c41e95b8796aa77bd3601209a15482979/wallet/src/lib.rs#L278

**Review by crusso — COMMENTED** [review]
Sorry, I complete lost track of this. Ping me on slack if you want me to respond quickly - I get totally lost in GH notifications.

**Comment by dfx-json** [comment]
hello and thank you for contribution. as this PR is quite outdated, I would like to close it. if you would still like to include this change, I encourage you to create a new PR. thank you for your understanding.

**Pull request closed without merging**